src/Chapter1.hs

```diff
@@ -541,8 +543,12 @@ value after "=" where the condition is true.
 
 Casual reminder about adding top-level type signatures for all functions :)
 -}
-
-mid x y z = error "mid: not implemented!"
+mid :: Int -> Int -> Int -> Int
+mid x y z
+    | ((x >= y) && (x <= z)) || ((x >= z) && (x <= y)) = x
+    | ((y >= x) && (y <= z)) || ((y >= z) && (y <= x)) = y
+    | ((z >= y) && (z <= x)) || ((z >= x) && (z <= y)) = z
+    | otherwise = x
```

Comment:
As we mentioned, the compiler in Haskell is very attentive to the exhaustive pattern-matching. And here it would warn you that `Pattern matching is not exhaustive`, as the guards have quite complicated logic, and the compiler won't be able to prove that it covers all the cases.

Because of that, you will need to add another guard – `| otherwise = z`, to tell the compiler, that your pattern matching is exhaustive 🙂


It looks like you can move all these 2 cases under a single otherwise and return z in here.
The implementation will be a bit shorter 🙂

Reply:
Hm, for some reason I can't figure out how to get the warning about non-exhaustive pattern matching neither in ghci nor in vscode. I only see that when tests are running

Reply:
Even when I try to to do this:

```haskell
f :: Maybe Int -> Int
f (Just a) = a
```

I get no warnings

Reply:
Non-exhaustive pattern-matching warnings are not enabled by default in GHC. They are enabled by the `-Wall` GHC flag. And you build your project with `cabal`, it passes such flag to GHC, so you see a warning. Apparently, VSCode plugin doesn't get it automatically from the project description :disappointed: